File: packages/upay-python/upay/resources/clients.py
"""
Recurso de Clientes
"""

import re
from typing import Optional, Dict, Any
from ..http import HttpClient


class ClientsResource:
    """Recurso para gerenciar Clientes"""
    
    def __init__(self, http: HttpClient):
        self.http = http
    
    def create(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Cria um novo cliente
        
        Args:
            data: Dados do cliente
                - name: Nome do cliente (obrigatório)
                - email: Email do cliente (obrigatório, válido)
                - document: CPF/CNPJ
                - phone: Telefone
                
        Returns:
            Cliente criado
        """
        # Validação básica
        if not data.get("name") or len(str(data["name"]).strip()) == 0:
            raise ValueError("Nome do cliente é obrigatório")
        
        if not data.get("email") or not self._is_valid_email(data["email"]):
            raise ValueError("Email inválido")
        
        return self.http.post("/clients", data)
    
    def list(
        self,
        page: Optional[int] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
        order_by: Optional[str] = None,
        order_direction: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Lista clientes
        
        Args:
            page: Número da página
            limit: Limite de itens por página
            cursor: Cursor para paginação
            order_by: Campo para ordenação
            order_direction: Direção da ordenação (ASC ou DESC)
            
        Returns:
            Dicionário com 'data' (lista) e 'pagination'
        """
        params = {
            "page": page,
            "limit": limit,
            "cursor": cursor,
            "orderBy": order_by,
            "orderDirection": order_direction,
        }
        
        response = self.http.get("/clients", params)
        
        # Mapear resposta: { message, clients, pagination } -> { data, pagination }
        return {
            "data": response.get("clients") or response.get("data") or [],
            "pagination": response.get("pagination") or {"total": 0, "page": 1, "limit": 10}
        }
    
    def get(self, client_id: str) -> Dict[str, Any]:
        """
        Obtém um cliente por ID
        
        Args:
            client_id: ID do cliente
            
        Returns:
            Cliente
        """
        if not client_id:
            raise ValueError("ID é obrigatório")
        
        return self.http.get(f"/clients/{client_id}")
    
    def update(self, client_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Atualiza um cliente
        
        Args:
            client_id: ID do cliente
            data: Dados para atualizar
            
        Returns:
            Cliente atualizado
        """
        if not client_id:
            raise ValueError("ID é obrigatório")
        
        if data.get("email") and not self._is_valid_email(data["email"]):
            raise ValueError("Email inválido")
        
        return self.http.patch(f"/clients/{client_id}", data)
    
    def _is_valid_email(self, email: str) -> bool:
        """Valida formato de email"""
        pattern = r'^[^\s@]+@[^\s@]+\.[^\s@]+$'
        return bool(re.match(pattern, email))

File: packages/upay-php/src/Resources/Clients.php
<?php

namespace Upay\Resources;

use Upay\HttpClient;

class Clients
{
    private HttpClient $http;
    
    public function __construct(HttpClient $http)
    {
        $this->http = $http;
    }
    
    public function create(array $data): array
    {
        // Validação básica
        if (!isset($data['name']) || !is_string($data['name']) || strlen(trim($data['name'])) === 0) {
            throw new \InvalidArgumentException('Nome do cliente é obrigatório');
        }
        
        if (!isset($data['email']) || !is_string($data['email']) || !$this->isValidEmail($data['email'])) {
            throw new \InvalidArgumentException('Email inválido');
        }
        
        return $this->http->post('/clients', $data);
    }
    
    public function list(?array $params = null): array
    {
        $response = $this->http->get('/clients', $params);
        
        return [
            'data' => $response['clients'] ?? $response['data'] ?? [],
            'pagination' => $response['pagination'] ?? ['total' => 0, 'page' => 1, 'limit' => 10]
        ];
    }
    
    public function get(string $id): array
    {
        if ($id === '' || trim($id) === '') {
            throw new \InvalidArgumentException('ID é obrigatório');
        }
        
        $encodedId = rawurlencode($id);
        return $this->http->get("/clients/{$encodedId}");
    }
    
    public function update(string $id, array $data): array
    {
        if ($id === '' || trim($id) === '') {
            throw new \InvalidArgumentException('ID é obrigatório');
        }
        
        // Validação de name (mesma lógica do create)
        if (isset($data['name'])) {
            if (!is_string($data['name']) || strlen(trim($data['name'])) === 0) {
                throw new \InvalidArgumentException('Nome do cliente é obrigatório');
            }
        }
        
        // Validação de email
        if (isset($data['email'])) {
            if (!is_string($data['email']) || !$this->isValidEmail($data['email'])) {
                throw new \InvalidArgumentException('Email inválido');
            }
        }
        
        $encodedId = rawurlencode($id);
        return $this->http->patch("/clients/{$encodedId}", $data);
    }
    
    private function isValidEmail(string $email): bool
    {
        return filter_var($email, FILTER_VALIDATE_EMAIL) !== false;
    }
}
